### PR TITLE
Mysql for Machine Group

### PR DESCRIPTION
**Pull request by RoshanKumarMicrosoft**

### CODE REVIEW

File: Tasks/MysqlDeploymentOnMachineGroupV1/task.json [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        "Release"
+    ],
+    "runsOn": [
+        "DeploymentGroup"
+    ],
+    "author": "Microsoft Corporation",
+    "version": {
+        "Major": 1,
+        "Minor": 0,
+        "Patch": 2
```
Comment (Ajay-MS) [inline Tasks/MysqlDeploymentOnMachineGroupV1/task.json]: Should minor version be 0 (zero)

File: Tasks/MysqlDeploymentOnMachineGroupV1/models/MysqlTaskParameter.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	private sqlInline: string;
+	private serverName: string;
+	private databaseName: string;
+	private sqlUserName: string;
+	private sqlPassword: string;
+	private sqlAdditionalArguments: string;
+
+    constructor() {
+        try {
+			this.taskNameSelector = tl.getInput('TaskNameSelector', true);
```
Comment (Ajay-MS) [inline Tasks/MysqlDeploymentOnMachineGroupV1/models/MysqlTaskParameter.ts]: Indentation

File: Tasks/MysqlDeploymentOnMachineGroupV1/models/MysqlTaskParameter.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            this.sqlFile = tl.getInput('SqlFile', false);
+            this.sqlInline = tl.getInput('SqlInline', false);
+            this.serverName = tl.getInput('ServerName', true);
+            this.databaseName = tl.getInput('DatabaseName', false);
```
Comment (Ajay-MS) [inline Tasks/MysqlDeploymentOnMachineGroupV1/models/MysqlTaskParameter.ts]: Why DB name is not a mandatory input.

File: Tasks/MysqlDeploymentOnMachineGroupV1/operations/ToolPathOperations.ts [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+     /**
+     * Get installed path of mysql for windows 
+     */
+    public async getInstalledPathOfMysqlForWindow(): Promise<string> {
+        let defer = Q.defer<string>();
+        // If user has installed 32 bit mysql client in 64 bit machine
+        this.getInstalledLocationFromPath("\\Software\\Wow6432Node\\MySQL AB").then((path) => {
+            task.debug('Window Wow6432 mysql executable path: '+path);
+            defer.resolve(path + "bin\\mysql.exe");
```
Comment (Ajay-MS) [inline Tasks/MysqlDeploymentOnMachineGroupV1/operations/ToolPathOperations.ts]: use path.join

File: Tasks/MysqlDeploymentOnMachineGroupV1/sql/MysqlClient.ts [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    private _toolPath: string;
+
+    constructor(mysqlTaskParameter: MysqlTaskParameter, toolPath: string) {
+        if (!mysqlTaskParameter) {
+            throw new Error(task.loc("MysqlTaskParameterCannotBeEmpty"));
+        }
+        if (!mysqlTaskParameter.getServerName() ||typeof mysqlTaskParameter.getServerName().valueOf() !== 'string') {
+            throw new Error(task.loc("MysqlServerNameCannotBeEmpty"));
+        }
+        if (!toolPath ||typeof toolPath.valueOf() !== 'string') {
```
Comment (Ajay-MS) [inline Tasks/MysqlDeploymentOnMachineGroupV1/sql/MysqlClient.ts]: spacing

File: Tasks/MysqlDeploymentOnMachineGroupV1/sql/MysqlClient.ts [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        this._mysqlTaskParameter = mysqlTaskParameter;
+        this._hostName = mysqlTaskParameter.getServerName();
+        this._toolPath = toolPath;
+    }
+
+    /**
+     * Get the connection argument for mysql
+     */
+    private _getArgumentString(): string{
+        let argumentString = "-h" + this._hostName + " -u" + this._mysqlTaskParameter.getSqlUserName() + " -p" + this._mysqlTaskParameter.getSqlPassword();
```
Comment (Ajay-MS) [inline Tasks/MysqlDeploymentOnMachineGroupV1/sql/MysqlClient.ts]: Check for password with special characters `~!@#$%^&*()_+-={}[]|\:;"'`

Comment (Ajay-MS) [reply Tasks/MysqlDeploymentOnMachineGroupV1/sql/MysqlClient.ts]: What is default timeout values for connection timeout and query timeout in MySQL?

File: Tasks/MysqlDeploymentOnMachineGroupV1/README.md [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+**Script file:** Use this option if you have a text file that has the necessary SQL statements to be executed on server. 
+ 
+ 
+- **Script path\*:** Full path of the script file on the automation agent or on a UNC path accessible to the automation agent like,  \\BudgetIT\DeployBuilds\script.sql. Also, predefined [system variables](https://msdn.microsoft.com/Library/vs/alm/Build/scripts/variables) like, $(agent.releaseDirectory) can also be used here. 
+ 
+ 
+**Inline script:** Use this option for running the Inline Script against the server.  
+ 
+ 
+- **Inline script\*:** Enter the actuall Script to be run.
```
[the remark this answers is not in the document]
Comment (RoshanKumarMicrosoft) [reply Tasks/MysqlDeploymentOnMachineGroupV1/README.md]: Done.

File: Tasks/MysqlDeploymentOnMachineGroupV1/task.json (hunk 1 above)
Comment (RoshanKumarMicrosoft) [reply Tasks/MysqlDeploymentOnMachineGroupV1/task.json]: Done.

File: Tasks/MysqlDeploymentOnMachineGroupV1/models/MysqlTaskParameter.ts (hunk 2 above)
Comment (RoshanKumarMicrosoft) [reply Tasks/MysqlDeploymentOnMachineGroupV1/models/MysqlTaskParameter.ts]: Done.

File: Tasks/MysqlDeploymentOnMachineGroupV1/operations/ToolPathOperations.ts (hunk 4 above)
Comment (RoshanKumarMicrosoft) [reply Tasks/MysqlDeploymentOnMachineGroupV1/operations/ToolPathOperations.ts]: It has escaped this is the reason not using path.join

File: Tasks/MysqlDeploymentOnMachineGroupV1/sql/MysqlClient.ts (hunk 5 above)
Comment (RoshanKumarMicrosoft) [reply Tasks/MysqlDeploymentOnMachineGroupV1/sql/MysqlClient.ts]: Done

File: Tasks/MysqlDeploymentOnMachineGroupV1/sql/MysqlClient.ts (hunk 6 above)
Comment (RoshanKumarMicrosoft) [reply Tasks/MysqlDeploymentOnMachineGroupV1/sql/MysqlClient.ts]: Default value is 0. Which means no timeout by default unless not explicitly set by user in mysql options for client or in mysql server setting options.

**Pull request merged**